File: dart/leet_code/2073.dart
/// Dart implementation of a hypothetical problem.
///
/// This module contains a class `Solution` with a method `timeRequiredToBuy`.
/// The `timeRequiredToBuy` method takes in two parameters:
/// - `tickets`: A list of integers representing the number of tickets each person has.
/// - `k`: An integer representing the index of the person in the queue.
///
/// The method returns an integer representing the time required for the person at index `k` to buy all their tickets.
/// Each person buys one ticket at a time, and it takes one unit of time to buy a ticket.
/// When a person has bought all their tickets, they leave the queue.
/// The person at index `k` continues to stay in the queue until they have bought all their tickets.
///
/// Error Handling:
/// - This method assumes that the input parameters are well-formed, i.e., `tickets` is a list of integers and `k` is an integer.
/// - If the input parameters are not well-formed, the behavior of the method is undefined.
library;

class Solution {
  /// Calculates the time required for the person at index `k` to buy all their tickets.
  ///
  /// This method uses a while loop to simulate the process of the people buying tickets.
  /// It iterates until the person at index `k` has bought all their tickets.
  /// In each iteration, it goes through the queue and each person buys one ticket if they still have tickets left.
  /// The time is incremented by one for each ticket bought.
  ///
  /// @param tickets The number of tickets each person has.
  /// @param k The index of the person in the queue.
  /// @return The time required for the person at index `k` to buy all their tickets.
  int timeRequiredToBuy(List<int> tickets, int k) {
    int time = 0;
    bool flag = true;
    while (flag) {
      for (int i = 0; i < tickets.length; i++) {
        // If the person at index `k` has bought all their tickets, stop the process.
        if (tickets[k] == 0) {
          flag = false;
          break;
        }
        // If the current person still has tickets left, they buy one ticket.
        if (tickets[i] > 0) {
          tickets[i]--;
          time++;
        }
      }
    }
    return time;
  }
}

File: csharp/leet_code/2073.cs
using System.Collections.Generic; // Required for List<>
using System.Linq; // Required for ToList()

/// <summary>
/// Solution class contains methods to solve the problem.
/// </summary>
public class Solution
{
    /// <summary>
    /// Calculates the time required to buy a ticket.
    /// </summary>
    /// <param name="tickets">An array of integers representing the number of tickets available at each counter.</param>
    /// <param name="k">An integer representing the index of the counter where the user is standing.</param>
    /// <returns>An integer representing the time required to buy a ticket.</returns>
    /// <exception cref="System.ArgumentOutOfRangeException">Thrown when 'k' is out of range of 'tickets' array.</exception>
    public int TimeRequiredToBuy(int[] tickets, int k)
    {
        // Convert the array to a list for easier manipulation
        List<int> ticketslist = tickets.ToList();

        // Check if 'k' is within the range of 'tickets' array
        if (k < 0 || k >= ticketslist.Count)
        {
            throw new System.ArgumentOutOfRangeException(nameof(k), "Index is out of range of 'tickets' array.");
        }

        // Counter to keep track of the time
        int counter = 0;

        // Loop until the tickets at the user's counter are not zero
        while (ticketslist[k] != 0)
        {
            // Iterate over all the counters
            for(int i = 0; i < ticketslist.Count; i++)
            {
                // Increment the time counter
                counter++;

                // Decrement the number of tickets at the current counter
                ticketslist[i]--;

                // If the tickets at the current counter are zero
                if (ticketslist[i] == 0)
                {
                    // If the user is at the current counter, break the loop
                    if(k == i)
                    {
                        break;
                    }

                    // Remove the current counter from the list
                    ticketslist.RemoveAt(i);

                    // If the current counter is before the user's counter, decrement the user's counter index
                    if (i < k)
                    {
                        k--;
                    }

                    // Decrement the loop counter to adjust for the removed counter
                    i--;
                }
            }
        }

        // Return the time counter
        return counter;
    }
}
